## 8장 수치형
데이터 분석에서 보통 수치형의 자료를 많이 다룹니다.

# 수치형 데이터로 변환
수치형 연산의 경우 같은 정수형이여도 부동 소수점 등의 자료가 있어 변환을 통해 통일을 해주어야 합니다.
예를들어 숙박 인원은 정수형이지만, 평균 숙박 인원은 소수점이 나올 수 있습니다.

In [1]:
type(40000/3)
# 자료형을 확인

float

In [2]:
int(40000/3)
# 정수형으로 변환

13333

In [3]:
float(40000/3)
# 부동 소수점형으로 변환

13333.333333333334

In [4]:
import pandas as pd
df = pd.DataFrame({'value' : [40000/3]})
df

value
0  13333.333333

In [5]:
df.dtypes
# 자료형 확인

value    float64
dtype: object

책에는 내용이 더 있지만, 기초 내용이여서 이정도만 적습니다.

# 대수화를 이용한 비선형 변화

선형은 말그대로 y = ax + b 처럼 일차원의 직선 모델로 가정하고, a, b 값을 찾는 것을 뜻합니다.

그러나 대부분의 사회현상은 비선형 모델입니다. 따라서 그중 많이 설명되는 것이 지수형태의 모델입니다.

그러나 지수형태 모델같은 비선형 모델은 연산에 많은 메모리가 필요합니다.

따라서 수학에서 이러한 지수형의 모델을 선형화 시키기 위한 도입한 방법이 바로 log입니다.

즉 로그를 통해 곱하기 연산을 더하기로 바꾸어 줄 수 있고, y = ax^n 의 모델을

log y = log a + n log x

log y 에 Y, log x 에 X를 대입하면

Y = log a + nX 비선형 모델도 선형화 시켜 해석이 가능합니다.

In [6]:
import numpy as np
# 파이썬에서 수학적인 해석을 위해선 numpy 라이브러리가 필요로 합니다.

In [7]:
reserve_tb = pd.read_csv('reserve.csv')
reserve_tb

reserve_id hotel_id customer_id     reserve_datetime checkin_date  \
0            r1     h_75         c_1  2016-03-06 13:09:42   2016-03-26   
1            r2    h_219         c_1  2016-07-16 23:39:55   2016-07-20   
2            r3    h_179         c_1  2016-09-24 10:03:17   2016-10-19   
3            r4    h_214         c_1  2017-03-08 03:20:10   2017-03-29   
4            r5     h_16         c_1  2017-09-05 19:50:37   2017-09-22   
...         ...      ...         ...                  ...          ...   
4025      r4026    h_129       c_999  2017-06-27 23:00:02   2017-07-10   
4026      r4027     h_97       c_999  2017-09-29 05:24:57   2017-10-09   
4027      r4028     h_27       c_999  2018-03-14 05:01:45   2018-04-02   
4028      r4029     h_48      c_1000  2016-04-16 15:20:17   2016-05-10   
4029      r4030    h_117      c_1000  2016-06-06 08:16:51   2016-07-06   

     checkin_time checkout_date  people_num  total_price  
0        10:00:00    2016-03-29           4        97200  
1        11:30:00    2016-07-21           2        20600  
2        09:00:00    2016-10-22           2        33600  
3        11:00:00    2017-03-30           4       194400  
4        10:30:00    2017-09-23           3        68100  
...           ...           ...         ...          ...  
4025     09:30:00    2017-07-11           2        16000  
4026     10:30:00    2017-10-10           2        41800  
4027     11:30:00    2018-04-04           2        74800  
4028     09:30:00    2016-05-13           4       540000  
4029     09:00:00    2016-07-09           1        44100  

[4030 rows x 9 columns]

In [8]:
reserve_tb['total_price_log'] = reserve_tb['total_price'].apply(lambda x : np.log10(x / 1000 + 1))
reserve_tb['total_price_log']

0       1.992111
1       1.334454
2       1.539076
3       2.290925
4       1.839478
          ...   
4025    1.230449
4026    1.631444
4027    1.879669
4028    2.733197
4029    1.654177
Name: total_price_log, Length: 4030, dtype: float64

# 범주화를 이용한 비선형 변화
키 성장을 예를 들겠습니다.

어릴때는 키가 빠르게 자라고, 점점 줄어들다가 어느 순간부터 정지할 것입니다.

위에 대수화는 계속 일정한 n 으로 자란다고 가정한 것입니다.

그러나 몇몇 경우에는 케이스에 따라 n을 다르게 적용 해야 합니다.

그러나 범주화를 자세하게 할 경우 표현해야할 경우 처리 비용이 늘고, 패턴또한 기하적으로 늘어날 것입니다.

이를 해결하기 위한 방법이 범주 플래그를 할때 if/elif/else 구문처럼

범주 수에서 범주 플래그를 하나 줄이는 것입니다.

예를들어 키 성장에서 연령을

0-9, 10-19, 20세 이상에서 0-9, 10-19 만 생각을 하면 됩니다.

두개 다 아닐경우 자동으로 20세 이상이기 때문입니다.

In [9]:
customer_tb = pd.read_csv('customer.csv')
customer_tb['age_rank'] = (np.floor(customer_tb['age']/10) * 10).astype('category')
customer_tb['age_rank']

C:\Users\User\anaconda3\lib\site-packages\pandas\io\formats\format.py:1405: FutureWarning: Index.ravel returning ndarray is deprecated; in a future version this will return a view on self.
  for val, m in zip(values.ravel(), mask.ravel())


0      40.0
1      30.0
2      40.0
3      40.0
4      30.0
       ... 
995    40.0
996    30.0
997    30.0
998    40.0
999    30.0
Name: age_rank, Length: 1000, dtype: category
Categories (7, float64): [20.0, 30.0, 40.0, 50.0, 60.0, 70.0, 80.0]

넘파이의 floor함수는 매개변수의 소수점 이하 자리수를 버리고, 매개변수 이하의 최대정수를 변환합니다.
예를들어 floor(15.4)는 15를 반환하고, floor(-3.4)는 -4를 반환합니다.
astype은 자료형을 변환하는 pandas의 함수입니다. 매개변수에 category를 지정하면, 판다스의 라이브러리의 범주형으로 변환됩니다.
여기서 우리가 가질 수 있는 값을 직접 지정을 해주려면, pandas.Categorical 함수를 이용합니다.

# 정규화
머신러닝에서 입력값이 가질 수 있는 범위가 크게 달라지면 큰 문제가 생깁니다.

분류 작업에서 데이터의 차이가 클 경우 큰 열의 영향이 강합니다.

연산을 통해 같은 위치에 있는 서로 다른 데이터에 대해 같은 영향력이 같도록 보정을 해줍니다.

In [10]:
from sklearn.preprocessing import StandardScaler

reserve_tb['people_num'] = reserve_tb['people_num'].astype(float)
# 정규화의 경우 필연적으로 소수점이 나올예정이므로 실수형태 자료로 자료를 변환해주어야 합니다.
ss = StandardScaler()

result = ss.fit_transform(reserve_tb[['people_num', 'total_price']])
result

array([[ 1.30070909, -0.05319398],
       [-0.48375308, -0.74782231],
       [-0.48375308, -0.62993499],
       ...,
       [-0.48375308, -0.25632289],
       [ 1.30070909,  3.96222931],
       [-1.37598416, -0.53471832]])

StandardScaler 함수는 앞서 말한 가우스 분포로의 정규화를 하는 함수입니다.

그 다음 정규화를 하는 객체 ss를 선언한 후, fit_transform()을 통해

people_num 과 total_price column에 대해 정규화를 하였습니다.

In [11]:
reserve_tb['people_normalized'] = [x[0] for x in result]
# result에서 x[0]은 사람 수에 대해 정규화 한 것 입니다.
reserve_tb['total_price_normalized'] = [x[1] for x in result]
# result에서 x[1]은 가격의 합에 대해 정규화 한 것 입니다.

reserve_tb[['people_normalized', 'total_price_normalized']]

people_normalized  total_price_normalized
0              1.300709               -0.053194
1             -0.483753               -0.747822
2             -0.483753               -0.629935
3              1.300709                0.828240
4              0.408478               -0.317080
...                 ...                     ...
4025          -0.483753               -0.789536
4026          -0.483753               -0.555575
4027          -0.483753               -0.256323
4028           1.300709                3.962229
4029          -1.375984               -0.534718

[4030 rows x 2 columns]

예를들어 어떤 정규화된 값이 1.3을 가진다하면, 가우스 적분 테이블이라고

그 값을 구하면 값의 상대적 위치를 알 수 있습니다.

# 예외값 제거
대부분의 값보다 극단적으로 크거나, 작은 값은 정규화 뿐만 아니라 예측 모델링을 할때에도 방해가 되므로, 전처리 과정에서도 제외를 하면

많은 도움이 됩니다.

그러나 극단적인 값을 고려해야할때에는 예외값이라도 제거하면 안됩니다.

대표적인 예로는 주가 분석에서 극단적인 값을 제거하는 것과 살리는 것은 수익률에서 많은 차이를 일으킵니다.

먼저 예외값을 제거하기위해선 일반적인 값과 극단적인 값을 구분을 하여야 합니다.

그중 많이 이용되는 방법이 정규화 인데, Z값의 크기가 3 이상인 경우에 대해서는 발생확률이 0.27%이므로

이러한 값들을 제거하는 것이 도움됩니다.

In [12]:
reserve_tb['total_price_normalized'].max()

7.2050372718932865

In [13]:
reserve_tb =reserve_tb[
    (abs(reserve_tb['total_price'] - np.mean(reserve_tb['total_price'])) / np.std(reserve_tb['total_price']) <= 3)].reset_index()
reserve_tb['total_price_normalized'].max()

2.8794796676976024

정규화는 (x - 평균)/표준편차 를 통해 진행되고, 이러한 과정을 통해 나온 값을 Z값이라 합니다.

여기서 보시면 예외값 제거를 하기전에는 Z값이 7.205 인것이 있지만, 예외값 제외를 통해

total_price_normalized의 최대값이 2.87인 것을 보았습니다.

# 주성분 분석을 이용한 차원 압축
예를들어 어떤 값이 z = f(x, y, a, b, c) 즉 x, y, a, b, c 5개의 값의 영향을 받습니다.

따라서 x, y, a, b, c 5개 변수를 살펴보아야 합니다.

그러나 여기서 어떤 식을 정의해 d = g(a, b, c)형태의 어느 식을 내가 모델링하고,

다시 z = f'(x, y, d) 형태의 식을 찾으면, 3개의 변수만을 살펴보아야 합니다.

이러한 과정을 차원 압축이라 합니다.

In [14]:
from sklearn.decomposition import PCA

pca = PCA(n_components=2)
pca

PCA(n_components=2)

In [15]:
production_tb = pd.read_csv('production.csv')

In [16]:
pca_values = pca.fit_transform(production_tb[['length', 'thickness']])

In [17]:
print('누적 기여율 : {0}'.format(sum(pca.explained_variance_ratio_)))
print('각 차원의 기여율 : {0}'.format(pca.explained_variance_ratio_))

누적 기여율 : 1.0
각 차원의 기여율 : [0.97897794 0.02102206]


In [18]:
pca_newvalues = pca.transform(production_tb[['length', 'thickness']])
pca_newvalues

array([[  76.96838157,  -13.38906936],
       [-112.11469337,   -8.24884796],
       [ -76.1994339 ,   11.19027127],
       ...,
       [  31.12100559,   15.48152593],
       [-117.87675543,    2.4361334 ],
       [   4.80243541,  -15.32174872]])

원리는 고유값 문제, 텐서 등 난이도 높은 수학이 등장하여 어렵습니다.

그냥 차원을 압축하여 데이터 가시화를 쉽게 만들거나, 새로운 차원에서 다른 발견을 할 수도 있다는 점을 기억하면 될 거 같습니다.

# 수치의 보완

데이터 발생 환경에 따라 데이터 결손은 흔합니다. 따라서 결손값에 전처리를 적용하기 전 어떠한 결손인지 확인해야 합니다.

결손은 크게 세 가지 종류가 있습ㄴ다.

MCAR : 완전 무작위 결손입니다. 예를들어 온도 측정에서 일정 확률로 실내 온도를 측정하는 온도 센서으 데이터가 파손하는 경우입니다.

MAR : 다른 항목 데이터에 의존한 결손입니다.예를 들어 온도가 높을수록 실내 온도를 측정하는 온도 센서의 데이터가 파손될 확률이 높아지는 경우입니다.

MNAR : 결손된 항목 데이터에 의존한 결손 입니다.예를들어 실내온도가 40도 넘으면 실내 온도를 측정하는 온도 센서의 데이터가 파손되는 경우입니다.

우선 결손 레코드 제거 방법이 있습니다.

In [19]:
production_miss_tb = pd.read_csv('production_missing_num.csv')
production_miss_tb

type      length           thickness  fault_flg
0      E  274.027383   40.24113135955541      False
1      D   86.319269  16.906714630016268      False
2      E  123.940388  1.0184619943950775      False
3      B  175.554886   16.41492419553766      False
4      B  244.934740  29.061080805480326      False
..   ...         ...                 ...        ...
995    C  363.214163   48.36948254254505      False
996    D  134.773797  26.861665493072845      False
997    B  231.174985   7.087471454014727      False
998    D   81.613510   5.716271403731385      False
999    C  202.010973  35.211646979555084       True

[1000 rows x 4 columns]

In [20]:
production_miss_tb.replace('None', np.nan, inplace=True)
production_miss_tb

type      length           thickness  fault_flg
0      E  274.027383   40.24113135955541      False
1      D   86.319269  16.906714630016268      False
2      E  123.940388  1.0184619943950775      False
3      B  175.554886   16.41492419553766      False
4      B  244.934740  29.061080805480326      False
..   ...         ...                 ...        ...
995    C  363.214163   48.36948254254505      False
996    D  134.773797  26.861665493072845      False
997    B  231.174985   7.087471454014727      False
998    D   81.613510   5.716271403731385      False
999    C  202.010973  35.211646979555084       True

[1000 rows x 4 columns]

In [21]:
production_miss_tb.dropna(subset=['thickness'], inplace=True)
production_miss_tb

type      length           thickness  fault_flg
0      E  274.027383   40.24113135955541      False
1      D   86.319269  16.906714630016268      False
2      E  123.940388  1.0184619943950775      False
3      B  175.554886   16.41492419553766      False
4      B  244.934740  29.061080805480326      False
..   ...         ...                 ...        ...
995    C  363.214163   48.36948254254505      False
996    D  134.773797  26.861665493072845      False
997    B  231.174985   7.087471454014727      False
998    D   81.613510   5.716271403731385      False
999    C  202.010973  35.211646979555084       True

[892 rows x 4 columns]

dropna()함수를 통해 NaN이 포함된 레코드를 제거하려고 합니다.

그러나 None은 문자열로 인식하여서 이를 NaN값으로 바꿔야 합니다.

따라서 replace() 함수를 통해 NaN으로 교환해주어야 합니다.

결손값을 보완하는 방법에는 특정 정수로 대체하는 방법도 있습니다.

In [22]:
production_miss_tb = pd.read_csv('production_missing_num.csv')
production_miss_tb

type      length           thickness  fault_flg
0      E  274.027383   40.24113135955541      False
1      D   86.319269  16.906714630016268      False
2      E  123.940388  1.0184619943950775      False
3      B  175.554886   16.41492419553766      False
4      B  244.934740  29.061080805480326      False
..   ...         ...                 ...        ...
995    C  363.214163   48.36948254254505      False
996    D  134.773797  26.861665493072845      False
997    B  231.174985   7.087471454014727      False
998    D   81.613510   5.716271403731385      False
999    C  202.010973  35.211646979555084       True

[1000 rows x 4 columns]

In [23]:
production_miss_tb.replace('None', np.nan, inplace=True)
production_miss_tb['thickness'].fillna(1, inplace = True)

또 다른 방법으로는 평균값 보완이 있습니다.

In [24]:
production_miss_tb = pd.read_csv('production_missing_num.csv')
production_miss_tb.replace('None', np.nan, inplace = True)
production_miss_tb['thickness'] = production_miss_tb['thickness'].astype('float64')

thickness_mean = production_miss_tb['thickness'].mean()

None 값에 평균값을 채워주기 위해 None을 NaN값으로 교체해주고,

mean()함수를 통해 평균을 구했습니다.

In [25]:
production_miss_tb['thickness'].fillna(thickness_mean, inplace = True)

또 다른 방법으론 PMM 방법이 있습니다.
(라이브 버리 설치문제로 아직 미구현)

In [39]:
from fancyimpute import IterativeImputer as MICE

In [40]:
production_miss_num = pd.read_csv('production_missing_num.csv')

In [41]:
production_miss_num.replace('None', np.nan, inplace=True)

production_miss_num['thickness'] = production_miss_num['thickness'].astype('float64')
production_miss_num['type'] = production_miss_num['type'].astype('category')
production_miss_num['fault_flg'] = production_miss_num['fault_flg'].astype('category')

In [42]:
production_dummy_flg = pd.get_dummies(production_miss_num[['type', 'fault_flg']], drop_first=True)

In [44]:
MICE.fit_transform()

TypeError: fit_transform() missing 1 required positional argument: 'X'